### DISCUSSION TITLE
How to format a phone numer

### DISCUSSION THREAD

**Discussion by Widar370** [microsoft-search/pnp-modern-search#3786] [Q&A]
I want to make the phone number in the person card easier to read, ie "12 34 56 78".
I asked ChatGPT and the reault was:
javascript:
methods: {
  formatPhoneNumber(phoneNumber) {
    return phoneNumber.replace(/(\d{2})(?=\d)/g, '$1 ');
  }
}
and then i html:
<div>{{ formatPhoneNumber(slot item @root.slots.MobilePhone) }}</div>

But this did not work. Any suggestions?

**Comment by wobba** [comment]
You cannot run random script in a handlebar template. You have to play with the available handlebar helper methods in e.g. https://github.com/helpers/handlebars-helpers?tab=readme-ov-file#categories and https://microsoft-search.github.io/pnp-modern-search/extensibility/templating/#handlebars-helpers

**Reply by Widar370** [reply]
I found this:
[{{phoneNumber}}](https://github.com/helpers/handlebars-helpers/blob/master/lib/number.js#L74)
Convert a string or number to a formatted phone number.

Params

num {Number|String}: The phone number to format, e.g. 8005551212
returns {Number}: Formatted phone number: (800) 555-1212

Is there a way to make this formatting using the locale setting?
In Norway the phonenumber 12345678 should be written 12 34 56 78.

**Reply by wobba** [reply]
Seems US only. You would need a library which knows formats for different locales as a custom function. And in Norway numbers starting with 4 or 9 should be formatted 999 99 999 ;)

**Reply by Widar370** [reply]
For your information: The main rule for formatting phone numbers in Norway changed in 2019. All numbers except those starting with 8, should be formatted 99 99 99 99. Numbers starting with 8 are formatted 899 99 999. Some other exceptions are numbers starting with 116 and numbers with three, four and five digits that are written without space.

**Reply by wobba** [reply]
> For your information: The main rule for formatting phone numbers in Norway changed in 2019. All numbers except those starting with 8, should be formatted 99 99 99 99. Numbers starting with 8 are formatted 899 99 999. Some other exceptions are numbers starting with 116 and numbers with three, four and five digits that are written without space.

Call me old :) I'm still a XXX XX XXX kind of guy 😅 and waiting for this to be written in some standard. Writing without spaces makes the easiest for copy/past at least.

Regardless, code has to be written to support formatting based on locale.

**Reply by Widar370** [reply]
It is written here: [nummerforskriften](https://lovdata.no/dokument/SF/forskrift/2004-02-16-426) § 16. Fra 22. mai 2019.
It is strange that a data program in 2024 is not able to show a phone number in a user frindly way and vise versa is not able to read user input with spaces. Likewise, using low-code, it should be easy to satisfy a user friendly result.

**Reply by kasperbolarsen** [reply]
But that IS possible.  Enter the phone number in the AD in the format you prefer,  and the display requirement is solved 😉 

As far as search being able to rewrite the query text to match some specific patterns,  then it is up to you handle that

**Reply by Widar370** [reply]
I know, everything is possible using some effort!
Since there exists some low-code:
{{phoneNumber}}
in handlebar-helpers, it would be nice to use it, but unfortenately this is only for US! That is, not user/programmer friendly!

**Reply by wobba** [reply]
> It is written here: [nummerforskriften](https://lovdata.no/dokument/SF/forskrift/2004-02-16-426) § 16. Fra 22. mai 2019. It is strange that a data program in 2024 is not able to show a phone number in a user frindly way and vise versa is not able to read user input with spaces. Likewise, using low-code, it should be easy to satisfy a user friendly result.

In order to format a number, all formatting rules for all countries has to be centralized. If you have a link to a .js lib doing this we can add a helper for it for sure. It has nothing to do with the year 2024 or not - someone has to do the consolidation work as there is no central agency distributing locale's for everything to all browsers/programs :)

**Comment by kasperbolarsen** [comment]
Phone numbers are a pain in the neck. The best solution I have seen is to have multiple phone number fields in the User Profiles. One for display format (+47 12 34 56 78) and another for search (12345678) and sometimes a 3rd for phone number extension search ( 5678 )
It is not pretty but it works
